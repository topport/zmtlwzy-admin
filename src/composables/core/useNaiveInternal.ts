import { darkTheme, lightTheme, GlobalTheme } from 'naive-ui';
import { useRootSetting } from '../setting/useRootSetting';

export function getNaiveCssVars(keyName: keyof GlobalTheme, isDark = false) {
  const _theme = isDark ? darkTheme : lightTheme;
  return _theme[keyName].self(_theme.common);
}

export function getAllNaiveCssVars(keyName: keyof GlobalTheme): Record<string, string> {
  const lightList = getNaiveCssVars(keyName);
  const darkList = getNaiveCssVars(keyName, true);

  const all = new Proxy(
    {},
    {
      get(_, propKey: string) {
        const match = propKey.match(/^(d_|dark_)(.+)$/);
        if (match) {
          return darkList[match[2]];
        }
        return lightList[propKey];
      },
    }
  );

  return Object.create(all);
}

export function getNaiveCssVarsRef(keyName: keyof GlobalTheme, propKey: string) {
  if (!getCurrentInstance()) return;
  const { getIsDarkMode } = useRootSetting();
  const lightList = getNaiveCssVars(keyName);
  const darkList = getNaiveCssVars(keyName, true);
  const cssVarRef = computed((): string => {
    return getIsDarkMode.value ? darkList[propKey] : lightList[propKey];
  });

  return cssVarRef;
}
