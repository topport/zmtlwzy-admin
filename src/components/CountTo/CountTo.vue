<template>
  <span :style="{ color }">
    {{ value }}
  </span>
</template>
<script setup lang="ts">
  import { useTransition, TransitionPresets } from '@vueuse/core';
  import { toWritableRef } from '/@/composables/utilities/toWritableRef';
  import { isNumber } from '/@/utils/is';

  import { propTypes } from '/@/utils/propTypes';

  const props = defineProps({
    startVal: propTypes.number.def(0),
    endVal: propTypes.number.def(9999),
    duration: propTypes.number.def(1200),
    isPlay: propTypes.bool.def(true),
    decimals: propTypes.number.def(0).validate((val: number) => val >= 0),
    prefix: propTypes.string.def(''),
    suffix: propTypes.string.def(''),
    separator: propTypes.string.def(','),
    decimal: propTypes.string.def('.'),
    delay: propTypes.number.def(0),
    color: propTypes.string,
    useEasing: propTypes.bool.def(true),
    transition: propTypes.string.def('easeOutExpo'),
  });

  const emit = defineEmits(['started', 'finished']);

  const source = toWritableRef(props, 'startVal');
  let outputValue = useTransition(source);

  const value = computed(() => formatNumber(outputValue.value));

  watch([() => props.startVal, () => props.endVal], () => {
    if (props.isPlay) {
      start();
    }
  });

  watch(
    () => props.isPlay,
    (val) => {
      val && start();
    },
    {
      immediate: true,
    }
  );

  function start() {
    run();
    source.value = props.endVal;
  }

  function run() {
    outputValue = useTransition(source, {
      delay: props.delay,
      duration: props.duration,
      onFinished: () => emit('finished'),
      onStarted: () => emit('started'),
      ...(props.useEasing ? { transition: TransitionPresets[props.transition] } : {}),
    });
  }

  function formatNumber(num: number | string) {
    if (!num && num !== 0) {
      return '';
    }
    const { decimals, decimal, separator, suffix, prefix } = props;
    num = Number(num).toFixed(decimals);
    num += '';

    const x = num.split('.');
    let x1 = x[0];
    const x2 = x.length > 1 ? decimal + x[1] : '';

    const rgx = /(\d+)(\d{3})/;
    if (separator && !isNumber(separator)) {
      while (rgx.test(x1)) {
        x1 = x1.replace(rgx, `$1${separator}$2`);
      }
    }
    return prefix + x1 + x2 + suffix;
  }
</script>
