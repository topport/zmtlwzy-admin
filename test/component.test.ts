import { mount } from '@vue/test-utils';
import { promiseTimeout } from '@vueuse/core';
import { describe, expect, it } from 'vitest';
import { NSwitch } from 'naive-ui';
import CountTo from '../src/components/CountTo/CountTo.vue';

describe('CountTo.vue', () => {
  it('should render', () => {
    const wrapper = mount(CountTo);
    expect(wrapper.text()).toContain('0');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render different', async () => {
    const wrapper = mount(CountTo, { props: { endVal: 233333, isPlay: false } });
    expect(wrapper.text()).toContain('0');
    await promiseTimeout(100);
    expect(wrapper.html()).toMatchInlineSnapshot('"<span>0</span>"');
    await wrapper.setProps({ isPlay: true });
    await promiseTimeout(1300);
    expect(wrapper.html()).toMatchInlineSnapshot('"<span>233,333</span>"');
  });
});

describe('NSwitch.vue', () => {
  it('should be render', async () => {
    const wrapper = mount(NSwitch, { slots: { checked: 'on', unchecked: 'off' } });
    expect(wrapper.find('.n-switch__checked').exists()).toBe(true);
    expect(wrapper.find('.n-switch__unchecked').exists()).toBe(true);
  });

  it('should be interactive', async () => {
    const wrapper = mount(NSwitch, { props: { defaultValue: true } });
    expect(wrapper.attributes('aria-checked')).toBe('true');
    await wrapper.trigger('click');
    expect(wrapper.attributes('aria-checked')).toBe('false');
  });
});
