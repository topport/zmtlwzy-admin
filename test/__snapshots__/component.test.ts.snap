// Vitest Snapshot v1

exports[`CountTo.vue > should render 1`] = `"<span>0</span>"`;
